instrumentation/opentelemetry_ecto/lib/opentelemetry_ecto.ex

```diff
@@ -51,62 +51,62 @@ defmodule OpentelemetryEcto do
       ) do
     # Doing all this even if the span isn't sampled so the sampler
     # could technically use the attributes to decide if it should sample or not
-
     total_time = measurements.total_time
     end_time = :opentelemetry.timestamp()
     start_time = end_time - total_time
-    database = repo.config()[:database]
 
-    url =
-      case repo.config()[:url] do
-        nil ->
-          # TODO: add port
-          URI.to_string(%URI{scheme: "ecto", host: repo.config()[:hostname]})
+    %{
+      hostname: hostname,
+      username: username,
+      database: database,
+      port: port
+    } =
+      [hostname: nil, username: nil, database: nil, port: nil]
+      |> Keyword.merge(repo.config())
+      |> Keyword.merge(Ecto.Repo.Supervisor.parse_url(repo.config()[:url] || ""))
+      |> Keyword.take([:hostname, :username, :database, :port])
+      |> Enum.into(%{})
 
-        url ->
-          url
-      end
+    url =
+      "ecto://#{hostname}:#{port}/#{database}"
+      |> URI.parse()
+      |> URI.to_string()
```

Comment:
What's the reason to parse the string and then output to a string again?

Reply:
To normalize the output

```elixir
hostname = "localhost"
port = 5432
atabase = "db"

"ecto://#{hostname}:#{port}/#{database}" |> URI.parse() |> URI.to_string()
 # => "ecto://myhost:5432/db"

# What if port is "" or nil
port = ""
"ecto://#{hostname}:#{port}/#{database}" |> URI.parse() |> URI.to_string()
# => "ecto://localhost/db"
```

Reply:
You don't need to start as a string then. Create a URI from the pieces with schema `ecto` and call `to_string`.